File: bot/add.py
from os import startfile
import discord
from discord.ext import commands

import json
import requests


class Admin(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @commands.command()
    async def add(self, ctx, url=''):
        if not url:
            return

        with open("../src/question.json", mode="r", encoding='utf-8') as f:
            qus = json.load(f)
            songAdd = False
        
        res = self.get_data(url)
        try:
            song = res['items'][0]
            lylic = song['lyrics'][0]['value'].replace('\r', '').split('\n\n')[0]
        except IndexError:
            await ctx.reply("この曲は追加できません")
            return
        
        if url.startswith('https://www.nicovideo.jp/watch/'):
            songid = url.replace('https://www.nicovideo.jp/watch/', '')
            if songid in qus["niconico"]:
                await ctx.reply("この曲は既に追加されています。")
                return
            qus["niconico"].append(songid)
            songAdd = True
        if url.startswith('http://nico.ms/'):
            songid = url.replace('http://nico.ms/', '')
            if songid in qus["niconico"]:
                await ctx.reply("この曲は既に追加されています。")
                return
            qus["niconico"].append(songid)
            songAdd = True

        if url.startswith('https://www.youtube.com/watch?v='):
            songid = url.replace('https://www.youtube.com/watch?v=', '')
            if songid in qus["youtube"]:
                await ctx.reply("この曲は既に追加されています。")
                return
            qus["youtube"].append(songid)
            songAdd = True
        if url.startswith('https://youtu.be/'):
            songid = url.replace('https://youtu.be/', '')
            if songid in qus["youtube"]:
                await ctx.reply("この曲は既に追加されています。")
                return
            qus["youtube"].append(songid)
            songAdd = True
        
        if songAdd == True:
            with open("../src/question.json", mode="w", encoding='utf-8') as f:
                WriteQu = json.dumps(qus, ensure_ascii=False, indent=2)
                f.write(WriteQu)
                f.close()
            songName = song["name"]
            await ctx.reply(f"楽曲 {songName} は正常に追加されました")



    def get_data(self, url):
        baseurl = "https://vocadb.net/api/songs"
        pyaload = {"query": url, "maxResults": "1", "fields": "Lyrics"}
        r = requests.get(baseurl, params=pyaload)
        return r.json()


def setup(bot):
    bot.add_cog(Admin(bot))

File: src/bot/commands/add.js
const { SlashCommandBuilder } = require('@discordjs/builders');
const fs = require('fs');
const request = require('request');

module.exports = {
	data: new SlashCommandBuilder().setName('add')
		.setDescription('Vocaleagueに楽曲を追加する')
		.addStringOption(
			option => option
				.setName('url')
				.setDescription('追加する楽曲url\n追加できない可能性有り')
				.setRequired(true)),
	async execute(interaction) {
		
		const url = interaction.options.getString('url')
		if (!url) {
			await interaction.reply("楽曲urlを指定してね\nusage: /add [ Youtube_url | NicoNico_url ]")
			return;
		}
		if (!url.startsWith('http')) {
			await interaction.reply("楽曲 __**url**__ を指定してね\nusage: /add [ Youtube_url | NicoNico_url ]")
			return;
		}
		
		request.get({
			uri: 'https://vocadb.net/api/songs',
			headers: {'Content-type': 'text/plain'},
			qs: {
				query: url,
				maxResults: 1,
				fields: 'Lyrics'
			},
			json: true
		}, async function(err, req, data){

			const song = data.items[0];

			try {
				const so = data['items'][0];
				const ly = so['lyrics'][0]['value'].replace('\r', '').split('\n\n')[0];
			} catch (err) {
				await interaction.reply(`${url}\nこの楽曲は追加できませんでした`);
				return;
			}
			if (url.startsWith('https://www.nicovideo.jp/watch/')) {
				await addToJson(url, 'https://www.nicovideo.jp/watch/', 'niconico', interaction, song);
			} else if (url.startsWith('http://nico.ms/')) {
				await addToJson(url, 'http://nico.ms/', 'niconico', interaction, song);
			} else if (url.startsWith('https://www.youtube.com/watch?v=')) {
				await addToJson(url, 'https://www.youtube.com/watch?v=', 'youtube', interaction, song);
			} else if (url.startsWith('https://youtu.be/')) {
				await addToJson(url, 'https://youtu.be/', 'youtube', interaction, song);
			}
			if (interaction.replied) {return;}
			await interaction.reply(`__**${data.items[0].name}**__ が正常に追加されました`);
			console.info(`${data.items[0].name} has been added.`)
			return;
		});
	},
};

async function addToJson (url, baseUrl, type, interaction, song) {
	const { youtube, niconico } = require('../../question.json')
	const songid = url.replace(baseUrl, '')
  const lylic = song.lyrics[0].value
  const answer = song.defaultName
	switch (type) {
		case 'niconico':
			const nicoid = songid.split('?')[0];
			for (n of niconico) {
				if (n.id === nicoid) {
					await interaction.reply('この曲は既に追加されてます');
					return;
				}
			}
			niconico.push({"id": nicoid,"title": answer,"lylic": lylic});
			break;
		case 'youtube':
			const youid = songid.split('&')[0];
			for (y of youtube) {
				if (y.id === youid) {
					await interaction.reply('この曲は既に追加されてます');
					return;
				}
			}
			youtube.push({"id": youid,"title": answer,"lylic": lylic});
			break;
		default:
			await interaction.reply('エラーが発生しました。')
			break;
	}
	const dataJ = {youtube: youtube, niconico: niconico};
	fs.writeFileSync('./src/question.json', JSON.stringify(dataJ, null, 2));
}
